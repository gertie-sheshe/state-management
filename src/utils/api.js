const API_URL = "http://localhost:3000";
const HEADERS = {
  Accept: "application/json",
  "Content-Type": "application/json",
};

export const apiRequest = async (url, options) => {
  const result = await fetch(url, options);
  const data = await result.json();

  return data;
};

export const getUsers = () => {
  return apiRequest(`${API_URL}/users`);
};

export const getTodos = () => {
  return apiRequest(`${API_URL}/todos`);
};

export const getUserTodos = (userId) => {
  return apiRequest(`${API_URL}/users/${userId}/todos`);
};

export const updateTodo = (todoId, data) => {
  return apiRequest(`${API_URL}/todos/${todoId}`, {
    method: "PUT",
    headers: HEADERS,
    body: JSON.stringify(data),
  });
};

export const createTodo = (data, userId) => {
  return apiRequest(`${API_URL}/users/${userId}/todos`, {
    method: "POST",
    headers: HEADERS,
    body: JSON.stringify(data),
  });
};
