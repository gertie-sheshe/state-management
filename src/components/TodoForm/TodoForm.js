import React, { useState } from "react";
import { useMutation, useQueryClient } from "react-query";
import { createTodo } from "../../utils/api";

function TodoForm({ userId }) {
  const [todo, setTodo] = useState("");
  const queryClient = useQueryClient();

  const addTodoMutation = useMutation(
    (options) => {
      const { data, id } = options;
      return createTodo(data, id);
    },
    {
      onSuccess: () => {
        queryClient.invalidateQueries("userTodos");
        setTodo("");
      },
    }
  );

  const handleSubmit = (e) => {
    e.preventDefault();
    const newTodo = {
      title: todo,
      completed: false,
    };

    addTodoMutation.mutate({ data: newTodo, id: userId });
  };

  const handleOnChange = (e) => {
    e.preventDefault();
    setTodo(e.target.value);
  };

  return (
    <div style={{ marginTop: "20px" }}>
      <form onSubmit={handleSubmit}>
        <label style={{ marginRight: "10px" }} htmlFor="todo">
          Add Todo:
        </label>
        <input
          disabled={!userId}
          id="todo"
          type="text"
          value={todo}
          onChange={handleOnChange}
        />
        <button disabled={!userId} style={{ marginLeft: "5px" }} type="submit">
          Add
        </button>
      </form>
      {addTodoMutation.error && <p>Error: {addTodoMutation.error}</p>}
      {addTodoMutation.isLoading && <p>Adding Todo...</p>}
    </div>
  );
}

export default TodoForm;
