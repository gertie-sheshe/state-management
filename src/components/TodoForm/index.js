import TodoForm from "./TodoForm";

export default TodoForm;

console.log("TODO FORM");
