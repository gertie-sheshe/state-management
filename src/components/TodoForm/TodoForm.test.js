import { render } from "../../tests/utils";
import TodoForm from "./TodoForm";

describe("TodoForm", () => {
  it("matches snapshot", () => {
    const { asFragment } = render(<TodoForm />);
    expect(asFragment()).toMatchSnapshot();
  });
});
