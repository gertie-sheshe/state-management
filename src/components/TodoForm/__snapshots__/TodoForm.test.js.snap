// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoForm matches snapshot 1`] = `
<DocumentFragment>
  <div
    style="margin-top: 20px;"
  >
    <form>
      <label
        for="todo"
        style="margin-right: 10px;"
      >
        Add Todo:
      </label>
      <input
        disabled=""
        id="todo"
        type="text"
        value=""
      />
      <button
        disabled=""
        style="margin-left: 5px;"
        type="submit"
      >
        Add
      </button>
    </form>
  </div>
</DocumentFragment>
`;
